packages/beacon-node/src/chain/bls/multithread/index.ts

```diff
@@ -28,6 +28,11 @@ import {
   jobItemSigSets,
   jobItemWorkReq,
 } from "./jobItem.js";
+import {asyncVerifyManySignatureSets} from "./verifyManySignatureSets.js";
+
+// defaultPoolSize should return core count - 1 to keep main thread on core. We
```

Comment:
I am wondering if we should always go with logical CPU count - 1 or have a max cap at some point, e.g. does it make sense to have pool size of 64 or even 128?

Same goes for any kind of worker pool we have

Reply:
I have it sized at number of cores. There is a fair amount of idle time on the main and network thread so the kernel is switching those out for bls in down times.

I need to update that note.

I tried with cores - 2 at first to keep main and network thread on at all times but there was performance degradation. Tried cores -1 also but found the better but similar. Best results were sizing the pool at the number of cores and letting the kernel do the management. Can fine tune further though I think

Reply:
But regarding my question, is there an upper bound to the pool where it just does not make sense to spawn more workers?

> I tried with cores - 2 at first to keep main and network thread on at all times but there was performance degradation.

I am surprised this makes a difference, because we kinda assume here that Lodestar is the only process running on the server which for most setups is not the case.

Reply:
I would assume more than the number of cores is generally bad.  So as an upper limit I tried to stick with number of cores.

The bls work is very math heavy and the assembly portions of the code (synchronous functions) stay on thread for most of the execution time.  Attempting to do more parallel calls than there are cores will net a reduction of overall performance because of the time swapping thread context (moving active thread to wait state while another thread is being executed by a core).  Similar to how we are moving the main thread and worker thread to wait while bls runs.  The kernel knows when a process is idle though and optimizes for this so that processes that are active get run and ones that are idle can wait for other threads to run.

There are some other nuances to that though. Multi-threading allows a core to have more than one "active" thread that can execute at any give time although this is not a standardized thing. And how/when the thread is "active" also is non-standard. Each manf. has a different core/thread ratio and different way of handling what is actually running assembly at a given time.  I would guess that we can tune the number of threads up slightly but how much will be empirical, and results will be specific to the machine they were tested with.

As for the part that the EL is also running on the box. This is a good point.  My evidence for setting the threadpool size was empirical.  Ran the same branch with a few different values on separate feature groups.  And was rough at best but the results were pretty obvious at the time however they did not get persisted anywhere other that me going "cores = threads is better" and I went with it.  There was a lot of variables floating around at the time so i just simply went with what looked best.  At this point, now that we have narrowed a lot down and have a much more stable set of factors it will be a good idea to do a more methodical approach to setting this value.

Also will be a good idea to include EL performance in the calculation for setting.  Not sure how to rate that side's efficiency but should be easy enough with a bit of digging.  Will def add this to the to-do list

Reply:
Thanks for the detailed answer

But again back to my question https://github.com/ChainSafe/lodestar/pull/6660#discussion_r1563993760, is there an upper bound we should set?

Let's assume a server has 256 cores, do we want spawn 256 workers or should we set a hardcoded upper bound where we know based on holesky, mainnet, and other data, that those workers would just be idle and consume memory for now good reason?

I know that the EF has these huge boxes, and it seems like a good practice to not spawn too many workers if we won't utilize them

Reply:
OHHH... I totally misunderstood what you meant.  Hmmm... The overhead for a thread that is unused is virtually nil so having them around is pretty low expense wise but let me think on that a bit.  Its common for a computer to have hundreds of (or more) threads at a given time.

 Idle capacity would just mean anything that comes into queue would have 0 wait time and there would never be all threads in active use.... I would have to dig a bit to see what downsides there are other than idle threads which are pretty low cost wise.

Reply:
I just noticed that at least for worker_threads, it's a noticeable overhead at startup, see issue during tests https://github.com/ChainSafe/lodestar/pull/5608 but I am not sue this matters. That's why I was asking because you have a better idea about this then I do

Reply:
> @nflaig as an example I found on OSX that cores = threads (ie 10 on my M1) but on our hzax41 boxes the threads = 2 x cores. Generally when requesting the "number of cores" or "how much multithreading" is available the number of logical cores is what is reported.

Also be aware of docker, often times I saw logical CPUs to equal physical cores there.

Reply:
> I just noticed that at least for worker_threads, it's a noticeable overhead at startup, see issue during tests #5608 but I am not sue this matters. That's why I was asking because you have a better idea about this then I do

worker_thread != kernel thread though.... for a worker thread a whole JS environment is spun up and that does take a considerable amount of time.  kernel threads are not the same... the time is not 0 but its very very very low (likely milliseconds).  Not sure how long though but I can tell you that starting this branch with libuv it goes to the next step almost immediately and starting with workers there is maybe a 10-15 second delay to start all the nodejs worker_threads

Found this on SO:
https://stackoverflow.com/questions/18274217/how-long-does-thread-creation-and-termination-take-under-windows

Reply:
> > @nflaig as an example I found on OSX that cores = threads (ie 10 on my M1) but on our hzax41 boxes the threads = 2 x cores. Generally when requesting the "number of cores" or "how much multithreading" is available the number of logical cores is what is reported.
> 
> Also be aware of docker, often times I saw logical CPUs to equal physical cores there.

I would guess, but need to confirm 100%, that docker would read the logical cpu count and report that to the container as the physical and logical count.  The docker runtime has a setting to throttle how many "CPU's" are available to the daemon.  I would assume they are using the soft/logical limit for that similar to how all software consumes the underlying system resources.

Reply:
> I would guess, but need to confirm 100%, that docker would read the logical cpu count and report that to the container as the physical and logical count. 

That was something I noticed when working on system metrics
https://github.com/ChainSafe/lodestar/blob/2dae605a9c83cfc349d9c89e85d6463006cf1c38/packages/beacon-node/src/monitoring/system.ts#L63-L67

Looking at the metrics from my server, it seems to be still the case

Reply:
> worker_thread != kernel thread though

yeah, that's what I thought as well, it might not matter for libuv threads but we should probably consider an upper bound for any other worker pools, although might not be relevant if we anyhow replace the BLS worker implementation, and it's not that problematic for keystore decryption as those are limited by number of keys already and will be removed once decryption is completed.